### PR TITLE
Array validation

### PR DESCRIPTION
**Pull request by int-index**
Implements `jArrayOf` and `jListOf`.

### CODE REVIEW

File: lib/Jijo/Validation.hs [hunk 1]
```diff
@@ -174,5 +178,19 @@ jValidateOptField fieldName vField o =
...
+  JValidation e (Vector.Vector a)
+jValidateElements vElement =
+  itraverse $ \i element ->
+    mapJValidationReport (scopeJValidationReport (JPSIndex i)) $
+    vElement element
+
 mapJValidationError :: (e -> e') -> JValidation e a -> JValidation e' a
 mapJValidationError f = mapJValidationReport (fmap f)
+
+-- Indexed 'traverse' using 'State'
```
Comment (neongreen) [inline lib/Jijo/Validation.hs]: Is it different from https://hackage.haskell.org/package/vector-0.12.0.3/docs/Data-Vector.html#v:imapM?

File: lib/Jijo/Validation.hs [hunk 2]
```diff
@@ -174,5 +178,19 @@ jValidateOptField fieldName vField o =
     mapJValidationReport (scopeJValidationReport (JPSField fieldName)) $
     vField field
 
+jValidateElements ::
```
Comment (neongreen) [inline lib/Jijo/Validation.hs:181]: Shouldn't there also be `jValidateElement` for just one element?

File: lib/Jijo/Validation.hs (hunk 1 above)
Comment (int-index) [reply lib/Jijo/Validation.hs]: Yes, this doesn't require `Monad` which `JValidation` does not implement.

File: lib/Jijo/Validation.hs (hunk 2 above)
Comment (int-index) [reply lib/Jijo/Validation.hs:181]: No, I don't think so. What would it do?

File: lib/Jijo/Validation.hs (hunk 1 above)
Comment (int-index) [reply lib/Jijo/Validation.hs]: I added a comment to clarify this.

Comment (neongreen) [reply lib/Jijo/Validation.hs]: https://github.com/haskell/vector/issues/132 – sigh.

File: lib/Jijo/Validation.hs (hunk 2 above)
Comment (neongreen) [reply lib/Jijo/Validation.hs:181]:
```haskell
 jValidateField ::                     jValidateElement ::
   Text ->                               Int ->
   (j -> JValidation e a) ->             (j -> JValidation e a) ->
   HashMap.HashMap Text j ->             Vector j ->
   JValidation e a                       JValidation e a
```

Comment (neongreen) [reply lib/Jijo/Validation.hs:181]: In fact, I also think there could be a `JArrayDefinition` that would be useful when working with JSON arrays containing different types. Aeson encodes tuples like that, but I assume there are APIs in the wild that also do that.

Comment (neongreen) [reply lib/Jijo/Validation.hs:181]: Whether it should be implemented in this PR or left as an open issue is a separate question though.

Comment (int-index) [reply lib/Jijo/Validation.hs:181]: OK, sounds reasonable. I haven't thought about heterogeneous sequences. Yes, let's do it separately when the need arises.

**Review by neongreen — APPROVED** [review]

**Pull request merged**